react/features/videosipgw/middleware.js

```diff
@@ -0,0 +1,182 @@
+
+import Logger from 'jitsi-meet-logger';
+import { CONFERENCE_WILL_JOIN } from '../base/conference';
+import {
+    SIP_GW_AVAILABILITY_CHANGED,
+    SIP_GW_INVITE_ROOMS
+} from './actionTypes';
+import {
+    JitsiConferenceEvents,
+    JitsiSIPVideoGWStatus
+} from '../base/lib-jitsi-meet';
+import { MiddlewareRegistry } from '../base/redux';
+import {
+    Notification,
+    showErrorNotification,
+    showNotification,
+    showWarningNotification
+} from '../notifications';
+
+const logger = Logger.getLogger(__filename);
+
+/**
+ * Middleware that captures conference video sip gw events and stores
+ * the global sip gw availability in redux or show appropriate notification
+ * for sip gw sessions.
+ * Captures invitation actions that create sip gw sessions or display
+ * appropriate error/warning notifications.
+ *
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case CONFERENCE_WILL_JOIN: {
+        const conference = getState()['features/base/conference'].joining;
+
+        conference.on(
+            JitsiConferenceEvents.VIDEO_SIP_GW_AVAILABILITY_CHANGED,
+            (...args) => dispatch(_availabilityChanged(...args)));
+        conference.on(
+            JitsiConferenceEvents.VIDEO_SIP_GW_SESSION_STATE_CHANGED,
+            event => {
+                const toDispatch = _sessionStateChanged(event);
+
+                // sessionStateChanged can decide there is nothing to dispatch
+                if (toDispatch) {
+                    dispatch(toDispatch);
+                }
+            });
+
+        break;
+    }
+    case SIP_GW_INVITE_ROOMS: {
+        const { status } = getState()['features/videosipgw'];
+
+        if (status === JitsiSIPVideoGWStatus.STATUS_UNDEFINED) {
+            dispatch(showErrorNotification({
+                descriptionKey: 'recording.unavailable',
+                descriptionArguments: {
+                    serviceName: '$t(videoSIPGW.serviceName)'
+                },
+                titleKey: 'videoSIPGW.unavailableTitle'
+            }));
+
+            return;
+        } else if (status === JitsiSIPVideoGWStatus.STATUS_BUSY) {
+            dispatch(showWarningNotification({
+                descriptionKey: 'videoSIPGW.busy',
+                titleKey: 'videoSIPGW.busyTitle'
+            }));
+
+            return;
+        } else if (status !== JitsiSIPVideoGWStatus.STATUS_AVAILABLE) {
+            logger.error(`Unknown sip videogw status ${status}`);
+
+            return;
+        }
+
+        for (const room of action.rooms) {
+            const { id: sipAddress, name: displayName } = room;
+
+            if (sipAddress && displayName) {
+                const newSession = action.conference
+                    .createVideoSIPGWSession(sipAddress, displayName);
+
+                if (newSession instanceof Error) {
```

Comment:
Why did you choose returning an error over other possible implementations, such as returning a failed promise?

Reply:
Cause we are creating a session and returning it, there is no network operation involved and no need for any promise. This was the initial implementation. 
But we can change the API of lib-jitsi-meet to be promised base, do you think its better that way?

Reply:
If it was the implementation that was already there then no I don't think you should change it right now so that I don't keep asking you to make additional changes. However, I do think it's better to be consistent with what returns because if a promise is always returned then errors can be handled in a catch.